src/tls/TLSClient.fst

```diff
@@ -0,0 +1,298 @@
+  let Correct (rem,hsm) = Handshake.parseHandshakeMessages (Some pv) (Some kex) payload in
+  let [(hs_msg,to_log)] = hsm in
+  IO.print_string ("Received HS("^(string_of_handshakeMessage hs_msg)^")\n");
+  let logged = handshakeMessageBytes (Some pv) hs_msg in
+  IO.print_string ("Logged message = Parsed message? ");
+  if (Platform.Bytes.equalBytes logged to_log) then IO.print_string "yes\n" else IO.print_string "no\n";
+  hs_msg,to_log 
+
+let sendRecordCallback callbacks pv ct msg str = 
+  let r = Record.makePacket ct pv msg in
+  let Correct _ = sendTcpPacket callbacks r in
+  IO.print_string ("Sending "^Content.ctToString ct^"Data("^str^")\n")
+
+let sendHSRecordCallback12 callbacks pv (m,b) = 
+  let str = string_of_handshakeMessage m in
+  sendRecordCallback callbacks pv Content.Handshake b str
+
+    
+let makePacket pv ct msg str = 
+  let r = Record.makePacket ct pv msg in
+  (*let _ = (match ct with
+  | Content.Application_data ->   IO.print_string ("Sending Data("^str^")\n")
+  | Content.Handshake ->   IO.print_string ("Sending HS("^str^")\n")
+  | Content.Change_cipher_spec ->   IO.print_string ("Sending CCS\n")
+  | Content.Alert ->   IO.print_string ("Sending Alert("^str^")\n")) in  *)
+  r
+  
+let rec really_check buf len = 
+    let lb = length buf in
+    if len = 0 then Correct buf
+    else if len = lb then Correct buf
+    else Error(AD_illegal_parameter,  "Unexpected buffer length")
+  
+let check_read header record pv =
+  match really_check header 5 with 
+  | Correct header -> (
+      match Record.parseHeader header with  (* contenType protocolVersion length *)
+      | Correct (ct,pv,len) -> (
+         match really_check record len with
+         | Correct payload  -> Correct (ct,pv,payload)
+         | Error e          -> Error e )
+      | Error e             -> Error e )
+  | Error e                 -> Error e
+
+let recvCCSRecordCallback callbacks pv = 
+  let Correct(Content.Change_cipher_spec,_,ccs) = recvRecordCallback callbacks pv in
+  IO.print_string "Received CCS\n";
+  ccs
+  
+let connect config callbacks =
+  let rid = new_region root in
+  let log = HandshakeLog.create #rid in
+
+  let ks, cr = KeySchedule.create #rid Client log in
+  let (ClientHello ch,chb) = Handshake.prepareClientHello config ks log None None in
+  let pv = ch.ch_protocol_version in 
+  let kex = TLSConstants.Kex_ECDHE in
+
+  sendHSRecordCallback12 callbacks pv (ClientHello ch,chb);
+
+  let (ServerHello(sh),shb) = recvHSRecordCallback12 callbacks pv kex log in
+  
+  let Correct (n,None) = Handshake.processServerHello config ks log None ch (ServerHello(sh),shb) in
+
+  let pv = n.n_protocol_version in
+  let cs = n.n_cipher_suite in
+  let CipherSuite kex sa ae = cs in
+  let ems = n.n_extensions.ne_extended_ms in
+  let sal = n.n_extensions.ne_signature_algorithms in
+
+  let (Certificate(sc),scb) = recvHSRecordCallback12 callbacks pv kex log in
```

Comment:
Nit: I think you wrote this code since the indentation is really good. Usually, parentheses are not needed for `let x, y = e in e'`